File: iosApp/Sources/Core/MusicCatalog.swift
import SwiftUI

/// 배경음악 카탈로그(순수 데이터) — Android `core.util.MusicCatalog` 의 Swift 포팅.
/// 트랙 정의(색·별모양·해금 업적)를 양 플랫폼이 동일하게 유지한다.
enum MusicCatalog {
    struct Track: Identifiable {
        let id: String
        let displayName: String
        let resName: String               // 번들 mp3 리소스명(확장자 제외)
        let colorHex: UInt32              // 별자리/다이얼 테마색(0xRRGGBB)
        let starType: Int                 // 다이얼 별 모양(StarStyle 0..8)
        let unlockAchievementId: String?  // nil = 기본 해금

        var color: Color { Color(hex: colorHex) }
    }

    /// 기본(처음부터 사용 가능) 음악 = 별의 속삭임.
    static let defaultId = "star_whisper"

    static let tracks: [Track] = [
        Track(id: "star_whisper",     displayName: "별의 속삭임", resName: "bgm_star_whisper",     colorHex: 0x6EE7B7, starType: 1, unlockAchievementId: nil),
        Track(id: "tiny_explorer",    displayName: "작은 탐험가", resName: "bgm_tiny_explorer",    colorHex: 0xFFD27F, starType: 4, unlockAchievementId: "first_step"),
        Track(id: "celestial_drift",  displayName: "천상의 표류", resName: "bgm_celestial_drift",  colorHex: 0x7FB7FF, starType: 0, unlockAchievementId: "storyteller"),
        Track(id: "cosmic_funk",      displayName: "코스믹 펑크", resName: "bgm_cosmic_funk",      colorHex: 0xFF9CC6, starType: 2, unlockAchievementId: "popular"),
        Track(id: "forgotten_galaxy", displayName: "잊혀진 은하", resName: "bgm_forgotten_galaxy", colorHex: 0xB89BFF, starType: 7, unlockAchievementId: "star_traveler"),
        Track(id: "nebula_garden",    displayName: "성운의 정원", resName: "bgm_nebula_garden",    colorHex: 0x9CE6A0, starType: 5, unlockAchievementId: "companion"),
    ]

    static func byId(_ id: String?) -> Track? {
        guard let id else { return nil }
        return tracks.first { $0.id == id }
    }

    static func index(of id: String?) -> Int {
        max(tracks.firstIndex { $0.id == id } ?? 0, 0)
    }

    static var defaultTrack: Track { byId(defaultId) ?? tracks[0] }
}

File: androidApp/src/main/java/com/chaminwoo/stary/core/util/MusicCatalog.kt
package com.chaminwoo.stary.core.util

/**
 * 배경음악 카탈로그(순수 데이터) — [MusicManager](재생) 와 음악 탭 UI 가 공유한다.
 *
 * - 음원은 `res/raw/` 에 평면으로 둔다(raw 하위 폴더는 Android 리소스로 인식 안 됨).
 * - 기본 음악([DEFAULT_ID]) 은 처음부터 해금. 나머지는 [unlockAchievementId] 업적 달성 시 해금.
 * - 색([colorArgb])·별 모양([starType]) 은 음악마다 달라 다이얼/별자리에 반영된다.
 *   (별자리 문양은 UI 쪽 MusicScreen 의 MUSIC_CONSTELLATIONS 가 id 로 매핑)
 */
object MusicCatalog {
    data class Track(
        val id: String,
        val displayName: String,
        val rawName: String,               // res/raw 리소스명(확장자 제외)
        val colorArgb: Long,               // 별자리/다이얼 테마색
        val starType: Int,                 // 다이얼 별 모양(StarStyle type 0..7)
        val unlockAchievementId: String?,  // null = 기본 해금
    )

    /** 기본(처음부터 사용 가능) 음악 = 별의 속삭임. */
    const val DEFAULT_ID = "star_whisper"

    val tracks: List<Track> = listOf(
        Track("star_whisper",    "별의 속삭임", "bgm_star_whisper",     0xFF6EE7B7, 1, null),
        Track("tiny_explorer",   "작은 탐험가", "bgm_tiny_explorer",    0xFFFFD27F, 4, "first_step"),
        Track("celestial_drift", "천상의 표류", "bgm_celestial_drift",  0xFF7FB7FF, 0, "storyteller"),
        Track("cosmic_funk",     "코스믹 펑크", "bgm_cosmic_funk",      0xFFFF9CC6, 2, "popular"),
        Track("forgotten_galaxy","잊혀진 은하", "bgm_forgotten_galaxy", 0xFFB89BFF, 7, "star_traveler"),
        Track("nebula_garden",   "성운의 정원", "bgm_nebula_garden",    0xFF9CE6A0, 5, "companion"),
    )

    fun byId(id: String?): Track? = tracks.firstOrNull { it.id == id }
    fun rawName(id: String?): String? = byId(id)?.rawName
    val default: Track get() = byId(DEFAULT_ID) ?: tracks.first()
    fun indexOf(id: String?): Int = tracks.indexOfFirst { it.id == id }.coerceAtLeast(0)
}
